src/librustc_mir/transform/validate.rs

```diff
@@ -49,6 +53,51 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         );
     }
 
+    fn check_ty_callable(&self, location: Location, ty: Ty<'tcx>) {
+        if let ty::FnPtr(..) | ty::FnDef(..) = ty.kind {
+            // We have a `FnPtr` or `FnDef` which is trivially safe to call.
+        } else if let Some(fn_once_trait) = self.tcx.lang_items().fn_once_trait() {
+            // FIXME(doctorn): call shims shouldn't reference unnormalized types, so
+            // this case should be removed.
```

Comment:
Did you try calling `normalize_erasing_regions` with `param_env.with_reveal_all()`?

Reply:
I tried `normalize_erasing_regions`, what does `param_env.with_reveal_all` actually do?

Reply:
I don't fully know. :/ It has something to do with how much stuff can be unfolded. For example, `with_reveal_all` will unfold `impl Trait` types to their underlying definition.

Reply:
It lets you see through `impl Trait` and specialization. Not sure that would do anything here though.

Reply:
The goal is to unfold `Self`.

Which `TyKind` variant even corresponds to `Self`?

Reply:
If I am not completely mistaken, `Self` is `TyKind::Param`, which is `substs[0]` when using `InternalSubsts::identity_for_item`.

`TyKind::Closure` should also be trivially safe to call here.

We still need the below branch in case someone implements `FnOnce` manually though. e.g. `Box<impl FnOnce>` is also callable and is represented as `TyKind::Adt` afaik.
```rust
fn foo(b: Box<impl FnOnce()>) {
    b()
}
```

Reply:
> TyKind::Closure should also be trivially safe to call here.

No. That is the type of *the environment of a closure*, which is not callable. It just implements a trait that permits some surface-level sugar. Miri will ICE when a `Call` terminator is a `Closure`.

Reply:
What you are saying makes sense on the HIR, but not any more on the MIR, which is more lower-level and where closure conversion has already happened.

Reply:
Ahhh, that's interesting :+1:

So my comment mentioning `Box<impl FnOnce>` is also irrelevant, as at this stage, the call of `b` should be represented as `TyKind::FnDef(def_id = FnOnce::call_once, substs = [Box<impl FnOnce()>, (())])` I think. Thanks :heart:

Reply:
>  as at this stage, the call of b should be represented as TyKind::FnDef(def_id = FnOnce::call_once, substs = [Box<impl FnOnce()>, (,)]) I think

Exactly. :)
(I used to have the exact same misconception.^^ The only reason I know this is that I found the relevant code in Miri at some point and realized it would actually ICE for `TyKind::Closure`... but closures work fine in Miri so something was clearly going on. This is actually great for Miri as it means we need to do no work at all to support closures; MIR construction has already done all the work for us.)